Improve hashtable key completion for type constrained variable assignments, nested hashtables and more

MartinGC94 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary
Fixes https://github.com/PowerShell/PowerShellEditorServices/issues/1852
Decided to make a full fix for all hashtable key completion scenarios I could think of:

- Fixes type inference for ConvertExpressions that use a PowerShell type that hasn't been loaded yet.
- Fix tab completion for constructor hashtables so it only completes properties if the class has a null constructor or if it's an interface.
- Fix tab completion for constructor hashtables that use a PowerShell type that hasn't been loaded yet.
- Fix tab completion for nested hashtables like: `Get-WinEvent -FilterHashtable @{SuppressHashFilter=@{<Tab>`
- Fix tab completion for hashtables that are in an array of arguments like: `ls | select Name,@{<Tab>}`
- Add the same property completion that exists for type conversions to typed variables command parameters and Property assignments
- Add completion for the module specification used in `using module @{<Tab>`

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

MartinGC94 committed:
Add completion for hashtable property definitions for typed variables and command parameters

dkaszews commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -7146,11 +7154,11 @@ internal static List<CompletionResult> CompleteHashtableKey(CompletionContext co
                         {
                             case "New-Object":
                                 var inferredType = AstTypeInference.InferTypeOf(commandAst, completionContext.TypeInferenceContext, TypeInferenceRuntimePermissions.AllowSafeEval);
-                                var result = new List<CompletionResult>();
```
Why change this? None of the variables exit the declaration scope as far as I can see.

MartinGC94 replied:
Because it was a little silly to have both a "result" and "results" variable when you can have just one. What's the drawback?

dkaszews replied:
Locality rule - variables should be declared at innermost scope, preferably at initialization. And you only ever used one of them, they never interacted. If I see a variable predeclared, I expect that the current scope will use assignment from an inner one:

```csharp
int i = 0;
if (condition) {
    i = ...;
    // use i
} else {
    i = bar();
    // use i
}
// Because author predeclared `i`, I will now look for its usage here
```

If you only wanted to unify the names, which is valid to me, you can do it without predeclaring.

MartinGC94 replied:
I cannot. results is used both inside and outside the if statements so the best location I can declare it is outside the if statements. It would probably work if I wrap it inside a silly if statement like `if (true)` but that's IMO a worse solution than just declaring it before all the if statements.

dkaszews replied:
Okay, I may have missed the outside usage

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -7195,18 +7203,29 @@ internal static List<CompletionResult> CompleteHashtableKey(CompletionContext co
+                        completionContext.TypeInferenceContext,
+                        new PSTypeName[] { new PSTypeName(binding.BoundParameters[parameterName].Parameter.Type) },
+                        results,
+                        $"{wordToComplete}*",
```
```suggestion
                        string.Concat(wordToComplete, "*"),
```

dkaszews replied:
Or just `wordToComplete + '*'`

iSazonov commented on test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:
```diff
@@ -227,44 +227,53 @@ switch ($x)
-            $C
+    context TypeConstructionWithHashtable {
+        BeforeAll {
+            class X {
```
X name a bit confuse me. Please use more informative name.

MartinGC94 committed:
Fixes type inference for ConvertExpressions that use a PowerShell type that hasn't been loaded yet.
Fix tab completion for constructor hashtables so it only completes properties if the class has a null constructor.
Fix tab completion for constructor hashtables that use a PowerShell type that hasn't been loaded yet.
Fix tab completion for nested hashtables like Get-WinEvent -FilterHashtable @{SuppressHashFilter=@{<Tab>
Fix tab completion for hashtables that are in an array of arguments like: ls | select Name,@{<Tab>}
Add the same property completion that exists for type conversions to typed variables command parameters and Property assignments

MartinGC94 committed:
Merge branch 'master' into AddCompletionForTypedVariables

MartinGC94 committed:
Add missing tests
Fix class name as requested by iSazonov
Add enumeration for the inferred types

MartinGC94 committed:
Exclude interfaces from the null constructor check

MartinGC94 committed:
Add completion for module specification keys in using module statement

MartinGC94 committed:
Merge branch 'master' into AddCompletionForTypedVariables

MartinGC94 committed:
Merge branch 'PowerShell:master' into AddCompletionForTypedVariables

iSazonov commented:
@SeeminglyScience Please review.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -7177,169 +7206,274 @@ internal static List<CompletionResult> CompleteStatementFlags(TokenKind kind, st
+                            object value;
+                            if (SafeExprEvaluator.TrySafeEval(pair.Item1, completionContext.ExecutionContext, out value))
+                            {
+                                var stringValue = value as string;
+                                if (stringValue is null)
+                                {
+                                    return null;
+                                }
```
Minor suggestion
```suggestion
                                if (value is string stringValue)
                                {
                                    return null;
                                }
```

daxian-dbw replied:
Oh, typo in my suggestion. I meant
```
if (value is not string stringValue)
{
    return null
}
```
But it's up to you.

iSazonov replied:
Why no? Sure?

MartinGC94 replied:
@iSazonov What? I have applied the updated suggestion if that's what you are asking.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -7177,169 +7206,274 @@ internal static List<CompletionResult> CompleteStatementFlags(TokenKind kind, st
+            while (parentAst is not null)
+            {
+                if (parentAst is HashtableAst parentTable)
+                {
```
I think you may want to check if `previousAst is HashtableAst` before running the `foreach` loop. Consider this scenario: `xxx = @{ abc = @(@{d<tab>}) }`

MartinGC94 replied:
I am checking it before the foreach loop. Either I'm misunderstanding something, or your editor tricked you somehow.

daxian-dbw replied:
I was asking about `previousAst`. I don't see it's checked before the foreach loop.
For some context, I suspect that you are expecting directly nested hashtable's, not something like this `@{ abc = @( @{d<tab>} ) }`. However, the code doesn't check if `previewAst` is a hashtable, and thus it's populating `nestedHashtableKeys` even it's not directly nested hashtable's.

MartinGC94 replied:
Ahh okay, my mistake then. I don't remember if I considered this scenario but I don't think it's necessarily wrong. Consider the following scenario:

```
class Level1
{
    [Level2[]] $Prop1
}
class Level2
{
    [string] $Prop2
}
[Level1] @{
    Prop1 = @(
        @{Prop2="Hello1"}
        @{Prop2="Hello2"}
    )
}
```
It would make sense to offer completions for the "Level2" properties when we can infer that the user is building an array of "Level2" objects for the "Level1" type.  
The current implementation isn't perfect though. It doesn't work when the types haven't been loaded yet and `Get-WinEvent -FilterHashtable @{SuppressHashFilter=@( @{<Tab>} )` offers completions even though arrays aren't valid for "SupressHashFilter".
I will have to think about how to solve this.

daxian-dbw replied:
That's a good example. It will be nice if this example can be put in code as comment.

This comment and https://github.com/PowerShell/PowerShell/pull/17660#discussion_r1176890611 below are about the same thing. Maybe think about if we can restrict the checks some more in this while loop.

MartinGC94 replied:
I've updated the check to only allow Asts I expect to see in relevant completion scenarios. I know some people frown on "goto" but it seems like the best solution here. Feel free to correct me if I'm wrong though.

daxian-dbw reviewed (commented):
Sorry for taking so long to review. Not finished yet, but want to share feedback so far.

MartinGC94 committed:
Merge branch 'PowerShell:master' into AddCompletionForTypedVariables

MartinGC94 commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6139,19 +6139,42 @@ private static void CompleteFormatViewByInferredType(CompletionContext context,
+                            }
+                        }
+                    }
+                    if (foundConstructor && !foundDefaultConstructor)
```
The hashtable syntax is valid as long as there's either no explicit constructors defined, or if one of the defined constructors take no arguments.  
This check handles both scenarios.

daxian-dbw replied:
Got it. When no constructor is defined, a default constructor will be created for it.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -6139,19 +6139,42 @@ private static void CompleteFormatViewByInferredType(CompletionContext context,
             {
-                if (typeNameUsed.Contains(psTypeName.Name))
+                if (!typeNameUsed.Add(psTypeName.Name)
+                    || (ignoreTypesWithoutDefaultConstructor && psTypeName.Type is not null && psTypeName.Type.GetConstructor(Type.EmptyTypes) is null && !psTypeName.Type.IsInterface))
```
Right, didn't expect that :) Thanks for checking.

MartinGC94 committed:
Apply Dongbo suggestion.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -7192,169 +7221,274 @@ internal static List<CompletionResult> CompleteStatementFlags(TokenKind kind, st
+                if (parentAst is CommandAst or ConvertExpressionAst or UsingStatementAst)
+                {
+                    break;
+                }
```
Shall we restrict the checks here some more? With this code, the following will trigger completion, but with the wrong completion results:
```
[System.Diagnostics.ProcessStartInfo] $pwd.CallMe(@(<tab>)
[ProcessStartInfo] $a = $myObj.CallSomething(@(<tab>))
```
Basically, the code currently ignores the ASTs in between the hashtable and the parent that is `assignment`, `convertexpresson`, `parameter` and etc.

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -7363,32 +7497,73 @@ internal static List<CompletionResult> CompleteHashtableKey(CompletionContext co
 
-                                return null;
+                    IEnumerable<PSTypeName> inferredTypes;
+                    if (hashtableIsNested)
```
By reading the code, it doesn't seem you are expecting a nested hashtable here. Did you mean to use `if (!hashtableIsNested)` here?

MartinGC94 committed:
Return early when encountering unexpected Ast

daxian-dbw reviewed (approved):
I'm fine the use of `goto LABEL` here to break out of the loop. It's not that confusing to get what it does by reading the code.

daxian-dbw commented:
@MartinGC94 Thanks for the contribution and I appreciate your patience.

andschwa commented:
This is awesome work!
